layout: Add missing support for some alignment keywords on absolutely positioned elements

taniishkaaa opened the pull request:
Adds support for the `self-start`, `self-end`, `left`, and `right` alignment keywords for absolutely positioned elements.


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #34282 
- [X] There are tests for these changes

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

taniishkaaa committed:
Add missing support for some alignment keywords on absolutely positioned elements

Signed-off-by: taniishkaaa <tanishkasingh2004@gmail.com>

Loirooriol commented on components/layout_2020/positioned.rs:
```diff
@@ -936,25 +951,89 @@ impl<'a> AbsoluteAxisSolver<'a> {
+                }
+            },
+            _ => Some(alignment_container.origin),
         }
```
In fact I would do something like

```rs
        assert_eq!(
            wrapper_writing_mode.is_horizontal(),
            abspos_writing_mode.is_horizontal(),
            "Mixed horizontal and vertical writing modes are not supported yet"
        );
        let self_value_matches_container = self.axis == AxisDirection::Block ||
            wrapper_writing_mode.is_bidi_ltr() == abspos_writing_mode.is_bidi_ltr();
        let resolved_value_after_safety = match value_after_safety {
            // https://drafts.csswg.org/css-align/#valdef-self-position-center
            AlignFlags::CENTER => AlignFlags::CENTER,
            // https://drafts.csswg.org/css-align/#valdef-self-position-self-start
            AlignFlags::SELF_START if self_value_matches_container() => AlignFlags::START,
            AlignFlags::SELF_START => AlignFlags::END,
            // https://drafts.csswg.org/css-align/#valdef-self-position-self-end
            AlignFlags::SELF_END if self_value_matches_container() => AlignFlags::END,
            AlignFlags::SELF_END => AlignFlags::START,
            // https://drafts.csswg.org/css-align/#valdef-justify-content-left
            AlignFlags::LEFT if wrapper_writing_mode.is_bidi_ltr() => AlignFlags::START,
            AlignFlags::LEFT => AlignFlags::END,
            // https://drafts.csswg.org/css-align/#valdef-justify-content-right
            AlignFlags::RIGHT if wrapper_writing_mode.is_bidi_ltr() => AlignFlags::END,
            AlignFlags::RIGH => AlignFlags::START,
            // https://drafts.csswg.org/css-align/#valdef-self-position-end
            // https://drafts.csswg.org/css-align/#valdef-self-position-flex-end
            AlignFlags::END | AlignFlags::FLEX_END => AlignFlags::END,
            // https://drafts.csswg.org/css-align/#valdef-self-position-start
            // https://drafts.csswg.org/css-align/#valdef-self-position-flex-start
            _ => AlignFlags::START,
        };
        let origin = alignment_container.origin +
            match resolved_value_after_safety {
                AlignFlags::END => alignment_container.length - margin_box_axis.length,
                AlignFlags::CENTER => (alignment_container.length - margin_box_axis.length) / 2,
                _ => return None,
            };
        Some(origin)
```

Loirooriol commented on components/layout_2020/positioned.rs:
```diff
@@ -936,25 +951,89 @@ impl<'a> AbsoluteAxisSolver<'a> {
         &self,
         margin_box_rect: LogicalRect<Au>,
         content_box_rect: &mut LogicalRect<Au>,
+        wrapper_writing_mode: WritingMode,
```
Avoid "wrapper". The right term is "alignment container", which is the "containing block" in this case. "container" may be clear enough, but otherwise it should be one of the 2 full terms, not "wrapper".

taniishkaaa commented:
There are a few unexpected fails, should I update test expectations if this is desired?

Loirooriol commented:
`/css/css-flexbox/abspos/position-absolute-001.html` seems suspicious. Other could be because we don't support vertical writing modes. Can you investigate?

taniishkaaa commented:
Sure, I'll look into this and keep you updated!

Loirooriol commented on components/layout_2020/positioned.rs:
```diff
@@ -928,33 +943,67 @@ impl<'a> AbsoluteAxisSolver<'a> {
         }
 
-        match value_after_safety {
-            AlignFlags::CENTER | AlignFlags::SPACE_AROUND | AlignFlags::SPACE_EVENLY => Some(
```
Sorry, I was wrong, content distribution alignments like `space-around` and `space-evenly` are needed when we are using the static position.
Removing them breaks `/css/css-flexbox/abspos/position-absolute-001.html`.

taniishkaaa replied:
No worries, I've added them back

Loirooriol committed:
Check the direction of the alignment container, nits, test expectations

In this case we need to check the direction of the static position
containing block, not the actual containing block:
```html
<!DOCTYPE html>
<div style="position: relative">
  <div style="display: flex; flex-direction: column; width: 100px; height: 100px; border: solid; direction: rtl">
    <div style="position: absolute; width: 20px; height: 20px; background: cyan; top: 20px; align-self: self-start"></div>
    <div style="position: absolute; width: 20px; height: 20px; background: magenta; bottom: 20px; align-self: self-end"></div>
  </div>
</div>
```

Signed-off-by: Oriol Brufau <obrufau@igalia.com>

Loirooriol commented on components/layout_2020/positioned.rs:
```diff
@@ -928,33 +943,69 @@ impl<'a> AbsoluteAxisSolver<'a> {
-            _ => None,
-        }
+        let self_value_matches_container = self.axis == AxisDirection::Block ||
+            container_writing_mode.is_bidi_ltr() == abspos_writing_mode.is_bidi_ltr();
```
I suspect we actually need to compare with `shared_fragment.original_parent_writing_mode` instead of `container_writing_mode` when both offsets are auto. And rather than passing 3 writing modes around, it may be better to store them in the `AbsoluteAxisSolver`.

But not sure, I think it's better to land #34426 first and then we can test.

Loirooriol replied:
This is the relevant testcase:

```html
<!DOCTYPE html>
<div style="position: relative">
  <div style="display: flex; flex-direction: column; width: 100px; height: 100px; border: solid; direction: rtl">
    <div style="position: absolute; width: 20px; height: 20px; background: cyan; top: 20px; align-self: self-start"></div>
    <div style="position: absolute; width: 20px; height: 20px; background: magenta; bottom: 20px; align-self: self-end"></div>
  </div>
</div>
```

I have rebased your patch and pushed a fix.

Loirooriol reviewed (commented):
All the test failures are because Servo lacks vertical writing mode, so they are expected.
